src/LinAlg/hiopKrylovSolver.hpp

```diff
@@ -0,0 +1,151 @@
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+// AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+// EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Additional BSD Notice
+// 1. This notice is required to be provided under our contract with the U.S. Department
+// of Energy (DOE). This work was produced at Lawrence Livermore National Laboratory under
+// Contract No. DE-AC52-07NA27344 with the DOE.
+// 2. Neither the United States Government nor Lawrence Livermore National Security, LLC
+// nor any of their employees, makes any warranty, express or implied, or assumes any
+// liability or responsibility for the accuracy, completeness, or usefulness of any
+// information, apparatus, product, or process disclosed, or represents that its use would
+// not infringe privately-owned rights.
+// 3. Also, reference herein to any specific commercial products, process, or services by
+// trade name, trademark, manufacturer or otherwise does not necessarily constitute or
+// imply its endorsement, recommendation, or favoring by the United States Government or
+// Lawrence Livermore National Security, LLC. The views and opinions of authors expressed
+// herein do not necessarily state or reflect those of the United States Government or
+// Lawrence Livermore National Security, LLC, and shall not be used for advertising or
+// product endorsement purposes.
+
+/* implements the Krylov iterative solver
+* @file hiopKrylovSolver.hpp
+* @ingroup LinearSolvers
+* @author Nai-Yuan Chiang <chiang7@lnnl.gov>, LNNL
+* @author Cosmin G. Petra <petra1@lnnl.gov>, LNNL
+*/
+
+#ifndef HIOP_KrylovSolver
+#define HIOP_KrylovSolver
+
+#include "hiopMatrix.hpp"
+#include "hiopMatrixDense.hpp"
+#include "hiopVector.hpp"
+
+#include "hiop_blasdefs.hpp"
+
+#include "hiopCppStdUtils.hpp"
+
+#include "hiopLinearOperator.hpp"
+#include "hiopNlpFormulation.hpp"
+
+namespace hiop
+{
+
+/**
+ * Base class for Krylov Solvers used by HiOp
+ */
+
+class hiopKrylovSolver
+{
+public:
+  hiopKrylovSolver(hiopNlpFormulation* nlp,
+                   hiopLinearOperator* A_opr = nullptr,
+                   hiopLinearOperator* Mleft_opr = nullptr,
+                   hiopLinearOperator* Mright_opr = nullptr,
+                   hiopVector* x0 = nullptr);
+  virtual ~hiopKrylovSolver();
+
+  /** Solves a linear system.
+   * param 'x' is on entry the right hand side(s) of the system to be solved. On
+   * exit is contains the solution(s).
+   */
+  virtual bool solve(hiopVector& x) = 0;
+
+  /** Set the initial guess for the Krylov solver */
+  virtual bool set_x0(hiopVector& x0);
```

Comment:
`const hiopVector& x`

maybe add a setter that set the initial point uniformly equal to a constant?